pkg/builder/kaniko/kaniko.go

```diff
@@ -0,0 +1,248 @@
+)
+
+type Kaniko struct {
+	k8sClientset kubernetes.Interface
+	k8sNamespace string
+}
+
+var _ builder.Builder = &Kaniko{}
+
+func NewKaniko(k8sClientset kubernetes.Interface, namespace string) *Kaniko {
+	return &Kaniko{
+		k8sClientset: k8sClientset,
+		k8sNamespace: namespace,
+	}
+}
+
+func (k *Kaniko) Build(ctx context.Context, b *builder.BuilderOptions) (logs string, err error) {
+	job, err := prepareJob(b)
+	if err != nil {
+		return "", fmt.Errorf("error preparing Job: %w", err)
+	}
+
+	fmt.Printf("Creating Job: %s\n", job.Name)
+
+	cJob, err := k.k8sClientset.BatchV1().Jobs(k.k8sNamespace).Create(ctx, job, metav1.CreateOptions{})
+	if err != nil {
+		return "", fmt.Errorf("error creating Job: %w", err)
+	}
+
+	fmt.Printf("Waiting for Job completion: %s\n", cJob.Name)
+
+	kJob, err := k.waitForJobCompletion(ctx, cJob)
+	if err != nil {
+		return "", fmt.Errorf("error waiting for Job completion: %w", err)
+	}
+
+	fmt.Printf("Getting Pod from Job: %s\n", kJob.Name)
+
+	pod, err := k.getFirstPodFromJob(ctx, kJob)
+	if err != nil {
+		return "", fmt.Errorf("error getting Pod from Job: %w", err)
+	}
+
+	fmt.Printf("Getting container logs from Pod: %s\n", pod.Name)
+
+	logs, err = k.getContainerLogs(ctx, pod)
+	if err != nil {
+		return "", fmt.Errorf("error getting container logs: %w", err)
+	}
+
+	fmt.Printf("Cleaning up Job: %s\n", kJob.Name)
+
+	if err := k.cleanup(ctx, kJob); err != nil {
+		return "", fmt.Errorf("error cleaning up: %w", err)
+	}
+
+	fmt.Printf("Build completed successfully\n")
+
+	return logs, nil
+}
+
+func (k *Kaniko) waitForJobCompletion(ctx context.Context, job *batchv1.Job) (*batchv1.Job, error) {
+	watcher, err := k.k8sClientset.BatchV1().Jobs(k.k8sNamespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", job.Name),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error watching Job: %w", err)
+	}
+	defer watcher.Stop()
+
+	for {
+		select {
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				return nil, fmt.Errorf("watch channel closed unexpectedly")
+			}
+
+			j, ok := event.Object.(*batchv1.Job)
+			if !ok {
+				continue
+			}
+
+			if j.Status.Succeeded > 0 || j.Status.Failed > 0 {
+				// Job completed (successfully or failed)
+				return j, nil
+			}
+		case <-ctx.Done():
+			return nil, fmt.Errorf("context cancelled")
+		}
+	}
+}
+
+func (k *Kaniko) getFirstPodFromJob(ctx context.Context, job *batchv1.Job) (*v1.Pod, error) {
+	// Assuming there's only one Pod template in the Job
+	podList, err := k.k8sClientset.CoreV1().Pods(k.k8sNamespace).List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("job-name=%s", job.Name),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error listing Pods: %w", err)
+	}
+
+	if len(podList.Items) == 0 {
+		return nil, fmt.Errorf("no Pods found for the Job: %s", job.Name)
+	}
+
+	return &podList.Items[0], nil
+}
+
+func (k *Kaniko) getContainerLogs(ctx context.Context, pod *v1.Pod) (string, error) {
```

Comment:
I'd remove all the "gets" from these personally, will read more beautifully to be k.ContainerLogs() etc